src/main/java/org/opensearch/security/user/User.java

```diff
@@ -272,13 +272,13 @@ public synchronized final Map<String, String> getCustomAttributesMap() {
         return attributes;
     }
 
-    public final void addSecurityRoles(final Collection<String> securityRoles) {
+    public synchronized final void addSecurityRoles(final Collection<String> securityRoles) {
         if (securityRoles != null && this.securityRoles != null) {
             this.securityRoles.addAll(securityRoles);
         }
     }
 
-    public final Set<String> getSecurityRoles() {
+    public synchronized final Set<String> getSecurityRoles() {
```

Comment:
If you want to protect concurrent access to these methods you need a single locking mechanism, these `synchronized` keywords work at the method level, not the class level.

Make an field object to handle the lock, e.g. `private final Object securityRolesLock = new Object();`  then do `lock(securityRolesLock) { ... }` inside of all places where the securityRoles collection is interacted with.

IMO it isn't clear if this will address the underlying issue because concurrent modification can happen outside these methods by iterating over the collection outside of these methods.

Reply:
Right now I'm looking into splitting `getSecurityRoles` and `getMappedRoles` because the currently logic in place intermingles the 2 which causes the issue. `getSecurityRoles` is called within `mapRoles` which can be problematic because after `mapRoles` is finished executing then `addSecurityRoles` is called within `PrivilegesEvaluator`.

What I'd like to see is one thread doing the mapping of the roles and other threads can read from its result if its already been computed. I'll try different ways to write an automated test for this change.

Reply:
This PR would make this issue occur with much less frequency, but obviously it would be best to ensure a situation like this can never happen